llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp

```diff
@@ -198,6 +198,22 @@ Instruction *InstCombinerImpl::visitMul(BinaryOperator &I) {
   if (SimplifyAssociativeOrCommutative(I))
     return &I;
 
+  // mul (sext (icmp eq x, 0)), x -> 0
```

Comment:
You should fold `mul (sext X), Y` into `select X, -Y, 0` instead of simplifying the whole expression.

Reply:
Example: https://github.com/llvm/llvm-project/blob/9a9aa41dea83039154601082b1aa2c56e35a5a17/llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp#L444-L449